lib/core/src/Cardano/Pool/Rank.hs

```diff
@@ -0,0 +1,379 @@
+-- in order to rank them by desirability,
+-- and we need to know the stake that the user wants to delegate.
+scorePools
+    :: Ord poolId
+    => RewardParams
+    -> Map poolId (RewardInfoPool, a)
+    -> Coin -- ^ Stake that the user wants to delegate
+    -> Map poolId (PoolScore, RewardInfoPool, a)
+scorePools params pools t
+    = Map.fromList $ zipWith doScore sortedByDesirability areTop
+  where
+    RewardParams{nOpt} = params
+    areTop = replicate nOpt True ++ repeat False
+
+    doScore (d, (pid, pool, a)) isTop = (pid, (score, pool, a))
+      where
+        score = PoolScore
+            { _nonMyopicMemberReward
+                = nonMyopicMemberReward params pool isTop t
+            , _desirability = d
+            }
+
+    sortedByDesirability
+        = L.sortOn (Down . fst)
+        . map (\(pid,(pool, a)) -> (desirability params pool, (pid, pool, a)))
+        $ Map.toList pools
+
+{-------------------------------------------------------------------------------
+    Redelegation warning
+-------------------------------------------------------------------------------}
+data RedelegationWarning
+    = AllGood
+    | TooFewBlocks
+    | OtherPoolsBetter
+    deriving (Eq, Show)
+
+-- FIXME: Adapt message to take care of previous epoch.
+
+instance Buildable RedelegationWarning where
+    build AllGood = build $ unwords
+        [ "The pool to which you had delegated your stake"
+        , "gives rewards within expectations."
+        ]
+    build TooFewBlocks = build $ unwords
+        [ "The pool to which you have delegated your stake"
+        , "may have not performed as expected,"
+        , "please check your delegation choice."
+        ]
+    build OtherPoolsBetter = build $ unwords
+        [ "Other pools may offer higher rewards,"
+        , "please check your delegation choice."
+        ]
+
+-- | Compute redelegation warning from current pool performance.
+--
+-- Note: This function uses the 'performanceEstimate' for the pool
+-- that we delegate to, but ignores this fields for the argument
+-- 'StakePoolsSummary'.
+redelegationWarning
+    :: EpochNo
+        -- ^ Epoch when delegation was made
+    -> (RewardInfoPool, Coin)
+        -- ^ ( Info about the pool that we delegate to
+        --   , absolute stake that we delegate )
+    -> StakePoolsSummary
+        -- ^ Current summary of all stake pools (for comparison)
+    -> EpochNo
+        -- ^ Current epoch
+    -> RedelegationWarning
+redelegationWarning timeOfDelegation (info,user) StakePoolsSummary{..} now
+    | (sigma <= 0.6 * s && p < 0.85) || (sigma > 0.6 * s && p < 0.9)
+        = TooFewBlocks
+    | getPercentage mr < getPercentage mrstar * w
+        = OtherPoolsBetter
+    | otherwise
+        = AllGood
+  where
+    sigma = getPercentage $ stakeRelative info
+    s = 1 / fromIntegral (nOpt rewardParams)
+    p = performanceEstimate info
+
+    mr = currentROS rewardParams info user
+    mrstar = maximum (mr:returns)
```

Comment:
Good idea to include MR for our pool in the maximum.

Reply:
The power of unit testing reveals that `maximum` requires a non-empty argument. :joy: